### ISSUE TITLE
[ML] CodeT5+ TSX Post-Processing Refinement at Stage 4 Exit

### ISSUE THREAD

**Issue by oscharko** [oscharko-dev/workspace-dev#943]
## Summary

Integrate a quantized CodeT5+ model (~60-220M parameters) at the exit of Stage 4 (`codegen.generate`) to refine generated TSX — improving MUI prop usage, sx styling accuracy, and component composition quality.

Ref: Epic #857 | Depends on: #940 (ML Asset Management Infrastructure)

---

## Motivation

Stage 4 generates TSX deterministically via `generateArtifactsStreaming()` in `src/parity/generator-core.ts`. While `validation-feedback.ts` catches TypeScript-level errors (type mismatches, missing imports), it cannot detect **semantic** issues:

- Wrong MUI component variant (e.g., `outlined` vs `contained` Button based on design intent)
- Suboptimal `sx` prop values that are technically valid but visually incorrect
- Missed component composition opportunities (e.g., a Card with header/content/actions pattern)
- Incorrect responsive breakpoint application

A code-aware model like CodeT5+ can catch these patterns because it understands code semantics, not just syntax.

**Abgrenzung zu #935:** Issue #935 (Component Mapping Hints) learns which Figma components map to which code components. This issue operates *after* mapping — refining the actual TSX output. Different scope, different model, different integration point.

---

### Model

- **Base:** CodeT5+ (Salesforce, https://arxiv.org/abs/2305.07922)
- **Size:** 60M (CodeT5+-small) to 220M (CodeT5+-base) parameters
- **Quantization:** INT4 (q4) — 30-100 MB on disk, 200-400 MB in RAM
- **Format:** ONNX (documented export via Hugging Face Optimum)
- **Fine-tuning:** On MUI v7 + React 19 code patterns (separate training pipeline)

### Integration Point

**File:** `src/job-engine/services/codegen-generate-service.ts`
**Location:** After `generateArtifactsStreaming()` completes for each screen

```
Existing flow:
  generateArtifactsStreaming() → yield StreamingArtifactEvent → write TSX files

Enhanced flow (ml_inference mode only):
  generateArtifactsStreaming() → yield base TSX →
    CodeT5Refiner.refine(baseTsx, screenIR, tokens) →
      if refinedTsx passes Stage 5 validation: use refined
      else: keep original base TSX (deterministic fallback)
```

### Safety Gate

The refined TSX MUST pass Stage 5 (`validate.project`) — lint, typecheck, build. If validation fails on the refined output, the system automatically falls back to the original deterministic output. This ensures ML refinement never degrades build quality.

### Performance Targets

| Metric | Target |
|---|---|
| Per-screen-file refinement | 100-500 ms |
| 10-screen project total | 1-5 seconds |
| Model load time (cold) | <5 seconds |
| RAM footprint | 200-400 MB |

---

## Node.js 22 Feasibility

- CodeT5+ has documented ONNX export via Hugging Face Optimum
- Runs via `onnxruntime-node` (same runtime as the element classifier)
- Tokenizer available via Transformers.js or bundled as ONNX operator
- All inference local, no network calls

---

## Input/Output Contract

**Input:**
- Raw TSX source (string) from deterministic generator
- `ScreenIR` context for the current screen (layout, element types, semantic hints)
- `DesignTokens` (palette, typography, spacing)

**Output:**
- Refined TSX source (string) with targeted improvements
- Diff metadata: list of changes made (for auditability)
- Confidence score per change

---

## Risks

| Risk | Mitigation |
|---|---|
| Model introduces subtle bugs | Stage 5 validation gate; confidence threshold per change |
| Hallucinated imports/components | Post-processing filter: only allow known MUI imports |
| Training data specificity | Must fine-tune on MUI v7 + React 19; generic CodeT5+ insufficient |
| RAM on 32 GB MacBook | 400 MB acceptable alongside element classifier (total ~700 MB) |

---

## arXiv References

- CodeT5+: A New Family of Open Code Large Language Models: https://arxiv.org/abs/2305.07922
- UniXcoder: Unified Cross-Modal Pre-training for Code Representation: https://arxiv.org/abs/2203.03850

---

## Acceptance Criteria

- [ ] CodeT5+ model (q4) loads via MlAssetManager (#940)
- [ ] Refinement runs per screen file after deterministic generation
- [ ] Refined output passes Stage 5 validation OR falls back to original
- [ ] Changes are auditable (diff metadata persisted as artifact)
- [ ] Only active when `llmCodegenMode === "ml_inference"` AND model available
- [ ] No regression in deterministic mode (zero code path changes when mode is `deterministic`)
- [ ] Total refinement time <5 seconds for a 10-screen project
- [ ] RAM stays within 400 MB ceiling for the refinement model

---

## Dependencies

- **Requires:** #940 (ML Asset Management Infrastructure)
- **Complements:** #923 (Candidate Rule Compiler — operates at rule level, this at code level)
- **Related:** #935 (Component Mapping Hints — operates at mapping level, this at output level)
- **Validation:** Reuses Stage 5 (`validate.project`) as quality gate

**Comment by oscharko** [comment]
Consolidated into umbrella epic #857 for staged execution. Reopen when the feedback/ML track is actively funded and scheduled.
